apisix/discovery/nacos.lua

```diff
@@ -0,0 +1,322 @@
+        local protocol = string_sub(url, 1, protocol_idx + 2)
+        local user_and_password = string_sub(url, protocol_idx + 3, auth_idx - 1)
+        local arr = ngx_re.split(user_and_password, ":")
+        if #arr == 2 then
+            username = arr[1]
+            password = arr[2]
+        end
+        local other = string_sub(url, auth_idx + 1)
+        url = protocol .. other
+    end
+
+    if local_conf.discovery.nacos.prefix then
+        url = url .. local_conf.discovery.nacos.prefix
+    end
+
+    if str_byte(url, #url) ~= str_byte("/") then
+        url = url .. "/"
+    end
+
+    return url, username, password
+end
+
+
+local function get_page_service(infos, base_uri, token_param, page_num)
+    --  TODO Hardcode:page size=100,will rewrite by spacewander after merge
+    local path = str_format(service_list_path, page_num, 100) .. token_param
+    local data, err = get_url(base_uri, path)
+    if err then
+        return data, err, path
+    end
+
+    for _, service_name in ipairs(data.doms) do
+        core.table.insert(infos, service_name)
+    end
+    return data, err, path
+end
+
+
+local function iter_and_add_service_info(infos, base_uri, token_param)
+    local data, err, path = get_page_service(infos, base_uri, token_param, 1)
+    if err then
+        log.error("get_url:", path, " err:", err)
+        return
+    end
+
+    --  TODO Hardcode:page size=100,will rewrite by spacewander after merge
+    local maxPage = math.ceil(data.count / 100)
+    if maxPage == 0 then
+        return
+    end
+
+    -- more than 1 page,continue fetch other pages
+    if maxPage > 1 then
+        for i = 2, maxPage do
+            get_page_service(infos, base_uri, token_param, i)
+        end
+    end
+end
+
+
+local function get_services(base_uri, token_param)
+    local infos = core.table.new(0, 0)
+    iter_and_add_service_info(infos, base_uri, token_param)
+    return infos
+end
+
+
+local function fetch_full_registry(premature)
+    if premature then
+        return
+    end
+
+    local up_apps = core.table.new(0, 0)
+    local base_uri, username, password = get_base_uri()
+    local token_param, err = get_token_param(base_uri, username, password)
+    if err then
+        log.error("get_token_param error:", err)
+        if not applications then
+            applications = up_apps
+        end
+        return
+    end
+
+    local infos = get_services(base_uri, token_param)
+    if #infos == 0 then
+        applications = up_apps
+        return
+    end
+
+    local data, err
+    for _, service_name in ipairs(infos) do
+        data, err = get_url(base_uri, instance_list_path .. service_name .. token_param)
+        if err then
+            log.error("get_url:", instance_list_path, " err:", err)
+            if not applications then
+                applications = up_apps
+            end
+            return
+        end
+
+        for _, host in ipairs(data.hosts) do
+            if tostring(host.valid) == 'true' and
```

Comment:
I set up the environment and debug it myself.
Here is the `host`: {\"enabled\":true,\"instanceHeartBeatTimeOut\":15000,\"ipDeleteTimeout\":30000,\"instanceHeartBeatInterval\":5000,\"metadata\":{\"preserved.register.source\":\"SPRING_CLOUD\"},\"port\":18001,\"serviceName\":\"DEFAULT_GROUP@@APISIX-NACOS\",\"instanceId\":\"192.168.65.3#18001#DEFAULT#DEFAULT_GROUP@@APISIX-NACOS\",\"healthy\":true,\"ip\":\"192.168.65.3\",\"ephemeral\":true,\"clusterName\":\"DEFAULT\",\"weight\":1}

There is not `valid` field. The response doesn't match their API doc: https://nacos.io/en-us/docs/open-api.html

I am not sure if it is because the latest Nacos has changed the API. Maybe we need to use the tagged nacos image `2.0.0` instead of the latest one?